zrml/swaps/src/lib.rs

```diff
@@ -292,6 +292,10 @@ mod pallet {
             let params = PoolExitWithExactAmountParams {
                 asset,
                 asset_amount: |asset_balance: BalanceOf<T>, total_supply: BalanceOf<T>| {
+                    let mul: BalanceOf<T> =
+                        bmul(total_supply.saturated_into(), T::MaxInRatio::get().saturated_into())?
+                            .saturated_into();
+                    ensure!(pool_amount <= mul, Error::<T>::MaxInRatio);
                     let asset_amount: BalanceOf<T> = crate::math::calc_single_out_given_pool_in(
                         asset_balance.saturated_into(),
                         Self::pool_weight_rslt(&pool, &asset)?,
```

Comment:
A little below after the caculation this part can be found:
```rust
ensure!(
    asset_amount
        <= bmul(
            asset_balance.saturated_into(),
            T::MaxOutRatio::get().saturated_into()
        )?
        .saturated_into(),
    Error::<T>::MaxOutRatio
);
```

Is this still necessary? (is also related to `pool_join_with_exact_pool_amount`)

Reply:
To summarize, all the `pool_join_with_*` and `pool_exit_with_*` extrinsics check that asset amounts swapped in or out are within the `MaxInRatio` or `MaxOutRatio` relative to the asset balance.

Furthermore, the `*_exact_asset_amount` extrinsics perform this check _before_ calling any of the methods from `math.rs`, which is useful, as they make sure that the functions from `math.rs` are not called with unallowed values. On the other hand, the `*_exact_pool_amount` extrinsics only perform this check _after_ the `math.rs` functions are called, which, in my opinion, draws their use into question.

My suggestion is to leave this in until we have completely audited the Balancer implementation. I'd say this PR is more about putting the important checks in rather than removing unnecessary stuff.